lib/aligned-lockable-buffer.ts

```diff
@@ -0,0 +1,66 @@
+import { getAlignedBuffer } from '@ronomon/direct-io';
+import RWMutex = require('rwmutex');
+
+export interface AlignedLockableBuffer extends Buffer {
+	alignment: number;
+	lock: () => Promise<() => void>;
+	rlock: () => Promise<() => void>;
+	slice: (start?: number, end?: number) => AlignedLockableBuffer;
+}
+
+function attachMutex(
+	buf: Buffer,
+	mutex: RWMutex,
+	alignment: number,
+): AlignedLockableBuffer {
+	const buffer = buf as AlignedLockableBuffer;
+	buffer.alignment = alignment;
+	buffer.lock = mutex.lock.bind(mutex);
+	buffer.rlock = mutex.rlock.bind(mutex);
+	const bufferSlice = buffer.slice.bind(buffer);
+	buffer.slice = (...args) => {
+		const slice = bufferSlice(...args);
+		return attachMutex(slice, mutex, alignment);
+	};
+	return buffer;
+}
+
+export function createBuffer(
+	size: number,
+	alignment: number,
+): AlignedLockableBuffer {
+	return attachMutex(
+		getAlignedBuffer(size, alignment),
+		new RWMutex(),
+		alignment,
+	);
+}
+
+export function isAlignedLockableBuffer(
+	buffer: Buffer,
+): buffer is AlignedLockableBuffer {
+	return (buffer as AlignedLockableBuffer).rlock !== undefined;
```

Comment:
You can do this type check without the cast using
```suggestion
	return 'rlock' in buffer;
```
or
```suggestion
	return 'rlock' in buffer && buffer.rlock !== undefined;
```

Reply:
updated